### ISSUE TITLE
Port to TypeScript

### ISSUE THREAD

**Issue by samreid** [phetsims/circuit-construction-kit-common#749]
After having a positive experience in porting bending-light to TypeScript in https://github.com/phetsims/bending-light/issues/401, it seems beneficial to port Circuit Construction Kit to TypeScript.  I was previously reluctant to do so, since AC was nearing publication and I didn't want to delay it at all.  However, we are awaiting some design decisions and QA, and I have a better understanding of the scope of effort for the port, I think there is time to do it now.  This will also help us through QA, if it helps prevent bugs.  I will also be able to benefit from it right away, since right after CCK AC publication, we plan to continue with phet-io instrumentation in #589 and #513.  This repo has 17000 LOC, so it will take longer than bending light.  Also, like PhET-iO, there is a cost associated with having too many sims "instrumented" before we have established all our patterns, but this seems like a good opportunity to help establish patterns.

UPDATE: As I've started this process, TypeScript has helped clarify so many confusing parts, especially around the Modified Nodal Analysis that it seems important for the long term maintainability of the sim, and it should be included in 1.0.  I will endeavor to do so without delaying the publication timeline.

**Comment by samreid** [comment]
Discovered bugs:

```js
    if ( this.zoomAnimation ) {
      const overflow = this.zoomAnimation.step( dt ); // always returns undefined
      if ( overflow > 0 ) {
        this.zoomAnimation = null;
      }
    }
```

**Comment by samreid** [comment]
I'm at a good check-in point for the TypeScript work.  It's taken 16.5 hours (for 18668 LOC).  Things are generally in good shape.  

* I was finding too many problems to write them all down, but I noted some here: https://github.com/phetsims/circuit-construction-kit-common/issues/749#issuecomment-943860478.  As far as I could tell, none were user-visible bugs, but all of them were maintainability concerns, and I'm glad they have been resolved.
* This also made it easier to refactor the modified nodal analysis code with confidence, which has been needed for a long time and will boost maintainability.
* I left about 25 options as `any`. Converting those were a drag because we don't use TypeScript in common code at the moment, and I'm hesitant to invest too much in the workaround of marking them in phet-types.d.ts (which doesn't import types correctly or define things in the right spot). 
* There remain 35 `@ts-ignore` and 45 `any` (other than the options `any` in the previous bullet point).  Some due to lack of TypeScript in common code, some due to complication in the sim code.  Some indicate potential bugs in the code that haven't been resolved.  I wrote a note in the TypeScript Quick Start Guide https://github.com/phetsims/chipper/blob/master/doc/typescript-quick-start.md not to refer to occurrences of `@ts-ignore` or `any` as an exemplar.

I'm not planning more work on this at the moment.  I published dev.36 here: https://phet-dev.colorado.edu/html/circuit-construction-kit-ac/1.0.0-dev.36/phet/circuit-construction-kit-ac_en_phet.html and will request testing in https://github.com/phetsims/circuit-construction-kit-common/issues/590.  Closing.
